### ISSUE TITLE
[Bug] Streaming issues with SillyTavern

### ISSUE THREAD

**Issue by FitzWM** [LostRuins/koboldcpp#357]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [*] I am running the latest code. Development is very rapid so there are no tagged versions as of now.
- [*] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [*] I [searched using keywords relevant to my issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/filtering-and-searching-issues-and-pull-requests) to make sure that I am creating a new issue that is not already open (or closed).
- [*] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new bug or useful enhancement to share.

# Expected Behavior

I am able to stream from KoboldCPP to SillyTavern with no interruptions or breaks, and I am able to stop generation and restart it without breaking anything.

# Current Behavior

To start off, I will say that I'm not sure if this is a SillyTavern bug or a KoboldCPP bug, but I figured I would report it here. I'm having three main issues when trying to use streaming with SillyTavern's KoboldAI API setting, and all three occur across multiple web browsers.

1. SillyTavern will "lose connection" with the API every so often. It doesn't actually lose connection at all, from what I can tell, as simply opening up the API menu in ST and clicking Connect will fix it every time. Refreshing the page also works. But until I do one of those things, ST will refuse to communicate with KoboldCPP.

2. If I interrupt generation mid-stream - because I can tell the response is going off the rails or whatever - there is a decent chance that something will break with the connection between ST and KoboldCPP. Once this happens, ST informs me that streaming isn't supported by my model, and I'm not able to generate anything more. I usually have to restart everything to fix this one: ST and KoboldCPP. 

3. ST doesn't always send the stopping sequences I have set up, meaning that KoboldCPP keeps on generating even when it clearly should be stopped. I haven't been able to figure out any pattern with this, but it seems to happen more often with swipes than new messages. I usually have to refresh the page to get it sending them properly again, although this doesn't always work. This one definitely feels like more of an ST issue than a KoboldCPP one, but I haven't had the problem with any of the other major backends, so I thought I'd include it.

# Environment and Context

Please provide detailed information about your computer setup. This is important in case the issue is not reproducible except for under certain specific conditions.

* Physical (or virtual) hardware you are using, e.g. for Linux:

CPU: AMD 5800X
RAM: 64 GB DDR4-4000
GPU: NVIDIA RTX 3080 Ti
SSD: Western Digital WD_BLACK SN850 1 TB

* Operating System, e.g. for Linux:

Windows 11 22H2 (22621.1992)

* SDK version, e.g. for Linux:

Python 3.10.6 (because xformers complains in other projects if I use other versions)

# Failure Information (for bugs)

Please help provide information about the failure if this is a bug. If it is not a bug, please remove the rest of this template.

# Steps to Reproduce

Sadly, it's kind of hard to provide steps, as the bug does seem to be pretty random. For example, the API disconnect thing will just happen sometimes, even if I just leave the tab open and don't interact at all. Other times, it will all go swimmingly and I can go through a dozen swipes before the thirteenth one triggers it.

# Failure Logs

Please include any relevant log snippets or files. If it works under one configuration but not under another, please provide logs for both configurations and their corresponding outputs so it is easy to see where behavior changes.

I typically get some variation of the following socket error when the bugs crop up. I did check my firewall and all the ports seem to be set up properly.

```
Generation Aborted
Gen[WinError 10053] An established connection was aborted by the software in your host machine
erating (301 / 300 tokens)Generate: The response could not be sent, maybe connection was terminated?

Time Taken - Processing:0.3s (315ms/T), Generation:6.1s (205ms/T), Total:6.5s (4.6T/s)----------------------------------------
Exception happened during processing of request from ('127.0.0.1', 40968)
Traceback (most recent call last):
  File "socketserver.py", line 316, in _handle_request_noblock
  File "socketserver.py", line 347, in process_request
  File "socketserver.py", line 360, in finish_request
  File "koboldcpp.py", line 313, in __call__
  File "http\server.py", line 647, in __init__
  File "socketserver.py", line 747, in __init__
  File "http\server.py", line 429, in handle
  File "http\server.py", line 395, in handle_one_request
  File "socket.py", line 669, in readinto
ConnectionAbortedError: [WinError 10053] An established connection was aborted by the software in your host machine
----------------------------------------
```

**Comment by LostRuins** [comment]
There's an SSE streaming fix in the latest version 1.38, please try it. As for the connection loss, I have not experienced that myself, could it be some sort of connection timeout? Try 1.38, If you request for fewer tokens does that help?

**Comment by duncannah** [comment]
The issue lies on when SillyTavern tries to fetch the version and model from the API, these two requests sometimes fail, telling SillyTavern either:
- The API is down (causing issue 1)
- Streaming isn't supported because it can't get the version (causing issue 2)
- Isn't sending stop sequences to the API, because it can't get the version (causing issue 3)

I had created an issue on the ST repo at SillyTavern/SillyTavern#595 before realizing this is probably an issue on KoboldCPP's side -- this is similar to #321 where the API dumps header info into the HTTP response

**Comment by FitzWM** [comment]
I just wanted to confirm that the above fix in 1.39.1 seems to have solved my issues completely so far. I'll definitely report back if I run into anything, though. Thanks, @duncannah !
